contracts/upgrader/src/contract.rs

```diff
@@ -0,0 +1,317 @@
+}
+
+fn current_version(env: &Env, contract_address: &Address) -> String {
+    env.invoke_contract(
+        contract_address,
+        &Symbol::new(env, VERSION),
+        soroban_sdk::vec![env],
+    )
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::contract::{Upgrader, UpgraderClient, MIGRATE, UPGRADE, VERSION};
+    use axelar_soroban_std::UpgradeableInterface;
+    use soroban_sdk::testutils::{Address as _, MockAuth, MockAuthInvoke};
+    use soroban_sdk::{contract, contracterror, contractimpl, contracttype, Address, Symbol};
+    use soroban_sdk::{BytesN, Env, String};
+
+    /// A simple contract to test the upgrader
+    #[contract]
+    struct DummyContract;
+
+    /// Dummy contract logic before upgrade
+    #[contractimpl]
+    impl UpgradeableInterface for DummyContract {
+        type Error = ContractError;
+
+        fn version(env: Env) -> String {
+            String::from_str(&env, "0.1.0")
+        }
+
+        fn upgrade(env: Env, new_wasm_hash: BytesN<32>) -> Result<(), ContractError> {
+            Self::owner(&env).require_auth();
+
+            env.deployer().update_current_contract_wasm(new_wasm_hash);
+            Ok(())
+        }
+    }
+
+    impl DummyContract {
+        fn owner(env: &Env) -> Address {
+            env.storage().instance().get(&DataKey::Owner).unwrap()
+        }
+    }
+
+    #[contracttype]
+    enum DataKey {
+        Data,
+        Owner,
+    }
+
+    #[contracterror]
+    enum ContractError {
+        SomeFailure = 1,
+    }
+
+    /// Dummy contract logic after upgrade is loaded into WASM_AFTER_UPGRADE
+    ///
+    /// #[contractimpl]
+    /// impl UpgradeableInterface for DummyContract {
+    ///     type Error = ContractError;
+    ///
+    ///     fn version(env: Env) -> String {
+    ///         String::from_str(&env, "0.2.0")
+    ///     }
+    ///
+    ///     fn upgrade(env: Env, new_wasm_hash: BytesN<32>) -> Result<(), ContractError> {
+    ///         Self::owner(&env).require_auth();
+    ///
+    ///         env.deployer().update_current_contract_wasm(new_wasm_hash);
+    ///         Ok(())
+    ///     }
+    /// }
+    ///
+    /// #[contractimpl]
+    /// impl DummyContract {
+    ///     pub fn migrate(env: Env, migration_data: String) {
+    ///         Self::owner(&env).require_auth();
+    ///
+    ///         env.storage()
+    ///             .instance()
+    ///             .set(&DataKey::Data, &migration_data);
+    ///     }
+    ///
+    ///     fn owner(env: &Env) -> Address {
+    ///         env.storage().instance().get(&DataKey::Owner).unwrap()
+    ///     }
+    /// }
+    const WASM_AFTER_UPGRADE: &[u8] = include_bytes!("testdata/dummy.wasm");
+
+    #[test]
+    fn upgrade_and_migrate_are_atomic() {
+        let env = Env::default();
+
+        let contract_address = env.register(DummyContract, ());
+        let upgrader_address = env.register(Upgrader, ());
+
+        let owner = set_owner(&env, &contract_address);
```

Comment:
The owner is being set in a separate function for `DummyContract`. Ideally, the ownership should be established in the `__constructor` function or transferred from the previous owner to the new owner using `transfer_ownership` function, gated with `owner.require_auth()`. Would this approach align with `UpgradeableInterface` or is it beyond the scope of this PR? I'm opting to be cautious.

Reply:
I can move setting the owner to the constructor, but I would like to keep the `DummyContract` as simple as I can, because I don't want the upgrader to make assumption about what's available beyond the upgrade, version and migrate functions. The actual upgrade and migration functions need to be tested in each contract themselves, again without making any assumptions about the upgrader